### PR TITLE
Phasing out obsolete Qt code.

### PR DESCRIPTION
**Pull request by themactep**

### CODE REVIEW

**Comment by themactep** [comment]
Ah, Qt5: 5.9.5, Qt5: 5.12.8... I came into the world too early. LOL. Anyway, I can delete this request or you can keep it till the good time comes.

File: src/ProgressWidget.cc [hunk 1]
```diff
@@ -4,6 +4,8 @@
 ProgressWidget::ProgressWidget(QWidget *parent)
 	:QWidget(parent)
 {
+	QElapsedTimer timer;
```
Comment (t-paul) [inline src/ProgressWidget.cc:7]: This looks unused?

**Review by t-paul — CHANGES_REQUESTED** [review]
Thanks for the PR, I think we do want to have it even though still supporting the older platforms (going forward this is now starting at Ubuntu 18.04).

I've added a dedicated `qt-obsolete.h` to centralize the switches for cases that are incompatible with old Qt versions.

Sorry for pushing to your `master` branch :smiley:.

**Comment by ChrisCoxArt** [comment]
Since Qt 5.9.9 is no longer available, and the OpenSCAD build doesn't work with the Qt versions available.... (which also require build script updates)
This change should be expedited.

**Comment by t-paul** [comment]
@ChrisCoxArt what do you mean? All official builds on `master` are green, so there's no panic needed at this point. This is 5.9.5 on Ubuntu 18.04 and 5.15.2 on Windows via MXE.

**Comment by ChrisCoxArt** [comment]
> @ChrisCoxArt what do you mean? All official builds on `master` are green, so there's no panic needed at this point. This is 5.9.5 on Ubuntu 18.04 and 5.15.2 on Windows via MXE.

That might work if you already downloaded the files.
But take a look at https://download.qt.io/official_releases/qt/
5.12 is the oldest version on there.  5.9.9 is no longer available.  Also, some of the file naming has changed from what is in the download/build scripts.  And, of course, OpenSCAD won't build with Qt 5.12 because of changes in Qt types and headers.
Seriously, are you people not doing clean builds often enough to catch these errors?

**Comment by themactep** [comment]
Chris, Qt 5.9.5 is the version supplied with Ubuntu 18.04 LTS (https://packages.ubuntu.com/bionic/qt5-default) which is actively used in the wild and will be supported by Canonical up to April 2023, and has the end-of-life date five more years past that, in 2028. Hence the need of support for older versions of Qt. Not a big deal, actually. I love the elegant macro solution Torsten came up with. I just need to get my s##t together, find time and rebase recent changes from origin master branch onto my fork and send a new PR for a clean patch. Will do it next week.

**Comment by themactep** [comment]
By the way, OpenSCAD compiles fine not only with Qt 5.12 but with Qt 5.15.2, the most recent version on Debian Sid. That is exactly why I see those pesky deprecation warnings.

**Comment by ChrisCoxArt** [comment]
> By the way, OpenSCAD compiles fine not only with Qt 5.12 but with Qt 5.15.2, the most recent version on Debian Sid. That is exactly why I see those pesky deprecation warnings.

Well, I'm trying to compile on MacOS, and no - OpenSCAD most definitely does not compile with Qt 5.12.
So far all I'm hearing from you guys is "it works fine for me, but we don't care about the rest of our builds."

**Comment by themactep** [comment]
See #3910.

**Comment by themactep** [comment]
> Well, I'm trying to compile on MacOS, and no - OpenSCAD most definitely does not compile with Qt 5.12.

Chris, if you did not realize, GitHub performs clean builds and reports developers about issues, if any. Here, results of builds for MacOS https://github.com/openscad/openscad/actions/workflows/macos.yml That is exactly what Torsten means telling you all official builds on master are green. They all compile just fine.

> So far all I'm hearing from you guys is "it works fine for me, but we don't care about the rest of our builds."

Maybe it is time to learn to read compilation log? "Does not compile" is not an error message.

**Comment by t-paul** [comment]
There might be a compilation issue, but if that's the case the only useful thing is to create a new ticket with the details. Adding unspecific complaints to an unrelated PR is not going to help.

Problem with macOS is that there's no member of the dev team with the needed hardware and Apple does not allow virtualization on other hardware. There's a special branch for the macOS topic - https://github.com/openscad/openscad/tree/qt5-upgrade-macos.

**Comment by ChrisCoxArt** [comment]
> > Well, I'm trying to compile on MacOS, and no - OpenSCAD most definitely does not compile with Qt 5.12.
> Chris, if you did not realize, GitHub performs clean builds and reports developers about issues, if any. Here, results of builds for MacOS https://github.com/openscad/openscad/actions/workflows/macos.yml That is exactly what Torsten means telling you all official builds on master are green. They all compile just fine.

And yet we know for a fact that a clean build cannot work, because a server you download source from has not had the specified files (or filenames) available since at least May of 2021.  So, no, the builds are not just fine - something is seriously broken in your build process when it fails to notice that required files are missing.

> > So far all I'm hearing from you guys is "it works fine for me, but we don't care about the rest of our builds."
> 
> Maybe it is time to learn to read compilation log? "Does not compile" is not an error message.

Or it time for you to stop being condescending to an experienced developer and start listening to fact that your build is fundamentally busted on at least one platform... and you failed to notice for several months.
Also, way to drive away engineers who want to contribute, but can't because your code has been broken for months.

Given that the errors are due to version differences in Qt versus what the source expects, and given that this issue does not specify a platform, and given that the errors are obvious if you do a download and build (ie: the download won't work, then the build won't work).... Anyone would expect that this issue would, in fact, cover the compile problems you obviously have with later versions of Qt, even if it doesn't solve the download script failures.

After updating to OpenSCAD mainline source earlier tonight, and downloading Qt5.12.11, here are the compile errors.
Should I expect an answer with more condescension and dismissal, or will you actually try to investigate the problems with your builds?

Consolidate compiler generated dependencies of target OpenSCAD
[  1%] Building CXX object CMakeFiles/OpenSCAD.dir/OpenSCAD_autogen/mocs_compilation.cpp.o
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/mocs_compilation.cpp:2:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/moc_AboutDialog.cpp:10:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/../../../src/AboutDialog.h:4:
In file included from /Volumes/Development/openscad/src/qtgettext.h:15:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/QString:1:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/qstring.h:49:
In file included from /usr/local/include/QtCore/qchar.h:43:
/usr/local/include/QtCore/qglobal.h:675:43: error: no template named 'is_arithmetic_v' in namespace 'std'; did you mean 'is_arithmetic'?
         typename = std::enable_if_t<std::is_arithmetic_v<T> && std::is_arithmetic_v<U> &&
                                     ~~~~~^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/type_traits:1183:50: note: 'is_arithmetic' declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS is_arithmetic
                                                 ^
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/mocs_compilation.cpp:2:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/moc_AboutDialog.cpp:10:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/../../../src/AboutDialog.h:4:
In file included from /Volumes/Development/openscad/src/qtgettext.h:15:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/QString:1:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/qstring.h:49:
In file included from /usr/local/include/QtCore/qchar.h:43:
/usr/local/include/QtCore/qglobal.h:675:70: error: no template named 'is_arithmetic_v' in namespace 'std'; did you mean 'is_arithmetic'?
         typename = std::enable_if_t<std::is_arithmetic_v<T> && std::is_arithmetic_v<U> &&
                                                                ~~~~~^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/type_traits:1183:50: note: 'is_arithmetic' declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS is_arithmetic
                                                 ^
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/mocs_compilation.cpp:2:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/moc_AboutDialog.cpp:10:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/../../../src/AboutDialog.h:4:
In file included from /Volumes/Development/openscad/src/qtgettext.h:15:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/QString:1:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/qstring.h:49:
In file included from /usr/local/include/QtCore/qchar.h:43:
/usr/local/include/QtCore/qglobal.h:675:87: error: expected '>'
         typename = std::enable_if_t<std::is_arithmetic_v<T> && std::is_arithmetic_v<U> &&
                                                                                      ^
/usr/local/include/QtCore/qglobal.h:675:37: note: to match this '<'
         typename = std::enable_if_t<std::is_arithmetic_v<T> && std::is_arithmetic_v<U> &&
                                    ^
/usr/local/include/QtCore/qglobal.h:675:65: error: expected ',' or '>' in template-parameter-list
         typename = std::enable_if_t<std::is_arithmetic_v<T> && std::is_arithmetic_v<U> &&
                                                                ^
/usr/local/include/QtCore/qglobal.h:676:66: error: expected unqualified-id
                                     std::is_floating_point_v<T> == std::is_floating_point_v<U> &&
                                                                 ^
/usr/local/include/QtCore/qglobal.h:687:35: error: no type named 'Promoted' in namespace 'QTypeTraits::detail'
using Promoted = typename detail::Promoted<T, U>::type;
                 ~~~~~~~~~~~~~~~~~^~~~~~~~
/usr/local/include/QtCore/qglobal.h:687:43: error: expected ';' after alias declaration
using Promoted = typename detail::Promoted<T, U>::type;
                                          ^
/usr/local/include/QtCore/qglobal.h:699:31: error: no template named 'Promoted' in namespace 'QTypeTraits'
constexpr inline QTypeTraits::Promoted<T, U> qMin(const T &a, const U &b)
                 ~~~~~~~~~~~~~^
/usr/local/include/QtCore/qglobal.h:701:28: error: no template named 'Promoted' in namespace 'QTypeTraits'
    using P = QTypeTraits::Promoted<T, U>;
              ~~~~~~~~~~~~~^
/usr/local/include/QtCore/qglobal.h:702:5: error: unknown type name 'P'
    P _a = a;
    ^
/usr/local/include/QtCore/qglobal.h:703:5: error: unknown type name 'P'
    P _b = b;
    ^
/usr/local/include/QtCore/qglobal.h:707:31: error: no template named 'Promoted' in namespace 'QTypeTraits'
constexpr inline QTypeTraits::Promoted<T, U> qMax(const T &a, const U &b)
                 ~~~~~~~~~~~~~^
/usr/local/include/QtCore/qglobal.h:709:28: error: no template named 'Promoted' in namespace 'QTypeTraits'
    using P = QTypeTraits::Promoted<T, U>;
              ~~~~~~~~~~~~~^
/usr/local/include/QtCore/qglobal.h:710:5: error: unknown type name 'P'
    P _a = a;
    ^
/usr/local/include/QtCore/qglobal.h:711:5: error: unknown type name 'P'
    P _b = b;
    ^
/usr/local/include/QtCore/qglobal.h:715:31: error: no template named 'Promoted' in namespace 'QTypeTraits'
constexpr inline QTypeTraits::Promoted<T, U> qBound(const T &min, const U &val, const T &max)
                 ~~~~~~~~~~~~~^
/usr/local/include/QtCore/qglobal.h:718:31: error: no template named 'Promoted' in namespace 'QTypeTraits'
constexpr inline QTypeTraits::Promoted<T, U> qBound(const T &min, const T &val, const U &max)
                 ~~~~~~~~~~~~~^
/usr/local/include/QtCore/qglobal.h:721:31: error: no template named 'Promoted' in namespace 'QTypeTraits'
constexpr inline QTypeTraits::Promoted<T, U> qBound(const U &min, const T &val, const T &max)
                 ~~~~~~~~~~~~~^
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/mocs_compilation.cpp:2:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/moc_AboutDialog.cpp:10:
In file included from /Volumes/Development/openscad/build/OpenSCAD_autogen/UVLADIE3JM/../../../src/AboutDialog.h:4:
In file included from /Volumes/Development/openscad/src/qtgettext.h:15:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/QString:1:
In file included from /usr/local/opt/qt@5/lib/QtCore.framework/Headers/qstring.h:49:
In file included from /usr/local/include/QtCore/qchar.h:43:
In file included from /usr/local/include/QtCore/qglobal.h:1336:
/usr/local/include/QtCore/qtypeinfo.h:58:46: error: no template named 'is_trivially_copyable_v' in namespace 'std'; did you mean 'is_trivially_copyable'?
inline constexpr bool qIsRelocatable =  std::is_trivially_copyable_v<T> && std::is_trivially_destructible_v<T>;
                                        ~~~~~^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/type_traits:3539:50: note: 'is_trivially_copyable' declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS is_trivially_copyable
                                                 ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [CMakeFiles/OpenSCAD.dir/OpenSCAD_autogen/mocs_compilation.cpp.o] Error 1
make[1]: *** [CMakeFiles/OpenSCAD.dir/all] Error 2
make: *** [all] Error 2

**Comment by themactep** [comment]
You get errors in Qt code, not OpenSCAD: `/usr/local/include/QtCore/qglobal.h:675:43: error: no template named 'is_arithmetic_v' in namespace 'std'; did you mean 'is_arithmetic'?` meaning your toolchain is not compatible. `is_arithmetic_v` requires C++17 https://en.cppreference.com/w/cpp/types/is_arithmetic 

And I don't even write in C, nor use MacOS.

**Comment by t-paul** [comment]
See #3915, as mentioned earlier, I'm not able to actually test this on a macOS system, but with some luck it works first go. Qt version is still available, but they moved it to `archive` folder. The download script tries the main location first and falls back trying the `archive` folder.

And again, this is a pull request completely unrelated to this issue, so I'm going to lock that. Please move further discussion to a dedicated ticket.

**Pull request closed without merging**